components/operations-controller/controllers/operation_controller.go

```diff
@@ -16,36 +16,428 @@ package controllers
 
 import (
 	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/kyma-incubator/compass/components/director/pkg/resource"
+	"github.com/kyma-incubator/compass/components/director/pkg/webhook"
+	"github.com/kyma-incubator/compass/components/operations-controller/internal/client"
+	"github.com/kyma-incubator/compass/components/operations-controller/internal/director"
+	"github.com/kyma-incubator/compass/components/operations-controller/internal/tenant"
+	web_hook "github.com/kyma-incubator/compass/components/operations-controller/internal/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"time"
+
+	"github.com/kyma-incubator/compass/components/director/pkg/graphql"
+
+	corev1 "k8s.io/api/core/v1"
 
 	"github.com/go-logr/logr"
-	"k8s.io/apimachinery/pkg/runtime"
 	ctrl "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 
-	operationsv1alpha1 "github.com/kyma-incubator/compass/components/operations-controller/api/v1alpha1"
+	"github.com/kyma-incubator/compass/components/operations-controller/api/v1alpha1"
+)
+
+var (
+	ErrReconciliationTimeoutReached = errors.New("reconciliation timeout reached")
+	ErrWebhookTimeoutReached        = errors.New("webhook timeout reached")
+	ErrFailedWebhookStatus          = errors.New("webhook operation has finished with failed status")
+	ErrUnsupportedWebhookMode       = errors.New("unsupported webhook mode")
 )
 
 // OperationReconciler reconciles a Operation object
 type OperationReconciler struct {
-	client.Client
-	Log    logr.Logger
-	Scheme *runtime.Scheme
+	k8sClient      client.Client
+	webhookConfig  *web_hook.Config
+	directorClient director.Client
+	webhookClient  web_hook.Client
+	log            logr.Logger
+}
+
+func NewOperationReconciler(k8sClient client.Client, webhookConfig *web_hook.Config, directorClient director.Client, webhookClient web_hook.Client, logger logr.Logger) *OperationReconciler {
+	return &OperationReconciler{
+		k8sClient:      k8sClient,
+		webhookConfig:  webhookConfig,
+		log:            logger,
+		directorClient: directorClient,
+		webhookClient:  webhookClient,
+	}
 }
 
 // +kubebuilder:rbac:groups=operations.compass,resources=operations,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=operations.compass,resources=operations/status,verbs=get;update;patch
 
 func (r *OperationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("operation", req.NamespacedName)
+	ctx := context.Background()
+	logger := r.log.WithValues("operation", req.NamespacedName)
+
+	operation, err := r.k8sClient.Get(ctx, req.NamespacedName)
+	if err != nil {
+		logger.Error(err, fmt.Sprintf("Unable to retrieve %s from API server", req.NamespacedName))
+		return ctrl.Result{}, nil
+	}
+
+	operation = setDefaultStatus(*operation)
+
+	webhookCount := len(operation.Spec.WebhookIDs)
+	if webhookCount == 0 || webhookCount > 1 {
```

Comment:
I think that executing an operation with 0 hooks should also work end to end. Imagine that for app X there are delete hooks and we delete it async. We also delete app Y async, there are no webhooks for it, but its delete app api that is async in the end so it should still work

Reply:
We also have this restriction on the Director side and it will not allow to schedule the operation and fail the mutation if there are no webhooks for an async mutation.
It sounds strange to me to want to delete an async application with no webhooks registered for such a request. If that's intentional, then the delete should be sync, and if it is not intentional and the webhook registration was actually forgotten, then that should be fixed by registering the necessary webhooks before attempting the async request.